File: Linear/Dynamic/Linked List/Singly Linked List/02_Inserting/Beginning/Inserting.js
// Define a node in the linked list
class Node {
    constructor(data) {
        this.data = data;
        this.next = null;
    }
}

// Function to insert a new node at the beginning of the list
function insertAtFront(head, x) {
    const newNode = new Node(x);
    newNode.next = head;
    return newNode;
}

function printList(head) {
    while (head !== null) {
        process.stdout.write(head.data.toString())
        if (head.next !== null)
            process.stdout.write(" -> ")
        head = head.next
    }
    console.log()
}

// Create the linked list: 2 -> 3 -> 4 -> 5
let head = new Node(2);
head.next = new Node(3);
head.next.next = new Node(4);
head.next.next.next = new Node(5);
const x = 1
console.log("Linked List before insertion");
printList(head);

// Insert a new node at the front of the list
head = insertAtFront(head, x);

console.log("Linked List after insertion");
printList(head);

File: Linear/Dynamic/Linked List/Doubly Linked List/02_Inserting/End/Inserting.java
class Node {
    int data;
    Node prev;
    Node next;

    Node(int data) {
        this.data = data;
        this.prev = null;
        this.next = null;
    }
}

class InsertingEnd {
    /*Function to insert a new node at
    the end of doubly linked list*/
    public static Node insertAtEnd(Node head, int x) {
        Node newNode = new Node(x);     // Create a new node

        /*If the Linked List is empty, make
        the new node as the head and return*/
        if (head == null)
        return newNode;

        // Store the head reference in a temporary variable
        Node last = head;

        // Traverse till the last node
        while (last.next != null) {
            last = last.next;
        }

        last.next = newNode;        // Set the next of last node to new node

        newNode.prev = last;        // Set prev of new node to last node

        return head;                // Return the head of the list
    }

    public static void printList(Node head) {
        while (head != null) {
            System.out.print(head.data);
            if (head.next != null)
            System.out.print(" <-> ");
            head = head.next;
        }

        System.out.println();
    }

    public static void main(String[] args) {
        // Create a hardcoded doubly linked list:
        // 1 <-> 2 <-> 3
        Node head = new Node(1);
        Node second = new Node(2);
        Node third = new Node(3);

        head.next = second;
        second.prev = head;
        second.next = third;
        third.prev = second;

        int x = 4;
        System.out.println("Linked List before insertion");
        printList(head);

        // Insert a new node at the end of the list
        head = insertAtEnd(head, x);
        
        System.out.println("Linked List after insertion");
        printList(head);
    }
}